### PR TITLE
Ability to assign car to driver

### PR DESCRIPTION
**Pull request by VitaliyKovtun**
Driver should have the ability to assign car for himself

It should be a custom action
`PATCH /drivers/assign_car`
the required parameter `car_id`

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -4,6 +4,10 @@
   post 'sign_up', to: 'registrations#create'
   post 'sign_in', to: 'authorizations#create'
 
+  resources :drivers do
+    patch 'assign_car'
```
Comment (SavenkoSy) [inline config/routes.rb]: drivers namespace, without `driver_id` in params, the driver is current_driver

File: lib/events/assign_car.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    end
+
+    private
+
+    def assign_car
+      Repositories::Driver.new.update(data[:driver_id], car_id: data[:car_id])
+    end
+
+    def validate!
+      validator = Validators::Driver.new(data)
```
Comment (SavenkoSy) [inline lib/events/assign_car.rb]: It's a bad way. We wanna just validate Isn't this car a busy by another driver

File: app/lib/validators/driver.rb [hunk 3]
```diff
@@ -5,5 +5,12 @@ class Driver
     relation :drivers
 
     validates :name, presence: true, uniqueness: true
+    validate :already_taken_car?
+
+    def already_taken_car?
```
Comment (SavenkoSy) [inline app/lib/validators/driver.rb]: validate this in event

File: lib/events/assign_car.rb (hunk 2 above)
Comment (VitaliyKovtun) [reply lib/events/assign_car.rb]: I guess, we need multiple validations: one for required fields and another one for related car id.

Comment (SavenkoSy) [reply lib/events/assign_car.rb]: `Validators::Driver` validate all field which requires to the object, but this event has only 1 attribute of this object. Do you see what I mean? This validator will be invalid all the time.

File: lib/events/assign_car.rb [hunk 4]
```diff
@@ -21,6 +21,12 @@ def assign_car
     def validate!
       validator = Validators::Driver.new(data)
       invalid!(validator.errors.messages) if validator.invalid?
+      invalid!('car_id': 'has already been taken') if already_taken_car?
+    end
+
+    def already_taken_car?
+      car_ids = Repositories::Driver.new.all.map { |driver| driver[:car_id] }
```
Comment (SavenkoSy) [inline lib/events/assign_car.rb]: `invalid!(I18n.error.message) if Repositories::Driver.new.by_car_id(data[:car_id])`

Comment (SavenkoSy) [reply lib/events/assign_car.rb]: the method in `Repositories::Driver`:
`def by_car_id(car_id)`
  `drivers.where(car_id: car_id).one`
`end`

File: lib/events/assign_car.rb [hunk 5]
```diff
@@ -19,14 +19,7 @@ def assign_car
...
     def validate!
-      validator = Validators::Driver.new(data)
-      invalid!(validator.errors.messages) if validator.invalid?
-      invalid!('car_id': 'has already been taken') if already_taken_car?
-    end
-
-    def already_taken_car?
-      car_ids = Repositories::Driver.new.all.map { |driver| driver[:car_id] }
-      car_ids.include?(data[:car_id])
+      invalid!('car_id': 'has already been taken') if Repositories::Driver.new.by_car_id(data[:car_id])
```
Comment (SavenkoSy) [inline lib/events/assign_car.rb]: I18n

Comment (VitaliyKovtun) [reply lib/events/assign_car.rb]: l19n

File: app/lib/repositories/feedback.rb [hunk 6]
```diff
@@ -8,6 +8,10 @@ def initialize
       super(ROM.env)
     end
 
+    def all
```
Comment (SavenkoSy) [inline app/lib/repositories/feedback.rb]: Do we use this anywhere?

File: spec/lib/events/assign_car_spec.rb [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    let(:data) do
+      Factory.structs[:driver].to_h.merge(
+        driver_id: driver.id,
+        car_id: car.id
+      )
+    end
+    let(:driver) { Factory[:driver] }
+    let(:car) { Factory[:car] }
+
+    it { is_expected.to change { drivers_repository.all.first } }
```
Comment (SavenkoSy) [inline spec/lib/events/assign_car_spec.rb]: `drivers_repository.by_id(driver.id).from(nil).to(car.id)`

File: spec/lib/events/create_feedback_spec.rb [hunk 8]
```diff
@@ -3,18 +3,33 @@
...
-        user_id: user.id,
-        trip_id: trip.id
-      )
+      Factory.structs[:feedback, user_id: user.id, trip_id: trip.id].to_h
     end
     let(:user) { Factory[:user] }
     let(:trip) { Factory[:trip] }
 
-    it { is_expected.to change { users_repository.all.count }.by(+1) }
+    it { is_expected.to change { feedbacks_repository.all.count }.by(+1) }
```
Comment (SavenkoSy) [inline spec/lib/events/create_feedback_spec.rb]: `feedbacks_repository.feedbacks.count`

File: app/lib/repositories/feedback.rb (hunk 6 above)
Comment (VitaliyKovtun) [reply app/lib/repositories/feedback.rb]: In rspec test.

Comment (SavenkoSy) [reply app/lib/repositories/feedback.rb]: delete it

File: spec/lib/events/create_feedback_spec.rb (hunk 8 above)
Comment (VitaliyKovtun) [reply spec/lib/events/create_feedback_spec.rb]: well, it's you created ".all" methods for each repository. but if that's non needed, I'll delete it all.

File: app/lib/repositories/feedback.rb (hunk 6 above)
Comment (VitaliyKovtun) [reply app/lib/repositories/feedback.rb]: we use .all methods in every spec tests. but if that's not needed, I'll delete all.

File: spec/lib/events/assign_car_spec.rb (hunk 7 above)
Comment (VitaliyKovtun) [reply spec/lib/events/assign_car_spec.rb]: that won't work

File: spec/lib/events/create_feedback_spec.rb (hunk 8 above)
Comment (SavenkoSy) [reply spec/lib/events/create_feedback_spec.rb]: It's not needed, really

File: app/lib/repositories/trip.rb [hunk 9]
```diff
@@ -8,6 +8,10 @@ def initialize
       super(ROM.env)
     end
 
+    def all
```
Comment (SavenkoSy) [inline app/lib/repositories/trip.rb]: remove this one

**Pull request merged**